### PR TITLE
layernom inplace

### PR DESCRIPTION
**Pull request by Krovatkin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63420
* #63419

### CODE REVIEW

**Comment by vadimkantorov** [comment]
would be nice to have inplace options for normalization layers as first-class: https://github.com/pytorch/pytorch/issues/26288

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 1]
```diff
@@ -48,7 +48,7 @@ namespace native {
 std::tuple<Tensor, Tensor, Tensor> mkldnn_layer_norm_last_index_weight_bias_f32(
     const Tensor& input,
     IntArrayRef normalized_shape, const Tensor& weight, const Tensor& bias,
-    double eps) {
+    double eps, bool inplace) {
```
Comment (eellison) [inline aten/src/ATen/native/mkldnn/Normalization.cpp:51]: typically the way that inplace is handled is that there is just a `Tensor& result` input, and the callers are responsible for initializing it as a new tensor or as the input

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 2]
```diff
@@ -470,7 +471,11 @@ const RegisterOperators BroadOpReg({
 const RegisterOperators MKLDNNLayerNormOpReg({
     torch::jit::Operator(
         "prim::MKLDNNLayerNorm(Tensor input, int[] normalized_shape, Tensor? weight=None, Tensor? bias=None, float eps=1e-05, bool cudnn_enable=True) -> Tensor",
-        MKLDNNLayerNormOp,
+        [](Stack* stack) { MKLDNNLayerNormOp(stack, false); },
+        AliasAnalysisKind::FROM_SCHEMA),
+    torch::jit::Operator(
+        "prim::MKLDNNLayerNorm_(Tensor input, int[] normalized_shape, Tensor? weight=None, Tensor? bias=None, float eps=1e-05, bool cudnn_enable=True) -> Tensor",
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:477]: `Tensor input` -> Tensor (a!) input`

File: test/jit/test_freezing.py [hunk 3]
```diff
@@ -1939,7 +1950,42 @@ def test_layernorm(self):
...
-                    sub_model = torch.nn.Sequential(param[0], layernorm)
+                    # to generate non inplace tests we extend the use of layernorm's input
+                    for inplace in (True, False):
+                        sub_model = torch.nn.Sequential(param[0], layernorm) if inplace else ResidualLayernorm(param[0], layernorm)
+                        sub_model.eval()
+                        mod = torch.jit.freeze(torch.jit.script(sub_model))
+                        self.run_pass("convert_frozen_ops_to_mkldnn", mod.graph)
+                        # if weight and bias are present and shape is the last dimension
+                        # we should convert `aten::layer_norm` to `prim::MKLDNNLayerNorm`
+                        if layernorm.elementwise_affine and len(param[1]) == 1:
```
Comment (eellison) [inline test/jit/test_freezing.py:1961]: why is `param[1]` == 1 a check ?

File: test/jit/test_freezing.py [hunk 4]
```diff
@@ -1939,7 +1950,42 @@ def test_layernorm(self):
...
+                            FileCheck().check_count("aten::to_dense", 1, exactly=True).run(mod.graph)
+                        self.assertTrue(torch.allclose(sub_model(param[2]), mod(param[2]), 1e-04, 1e-04))
+
+    @skipIfNoTorchVision
+    def test_layernorm2(self):
+        with set_default_dtype(torch.float):
+            model = torchvision.models.resnet18()
+
+
+            class ResidualLayernorm(torch.nn.Module):
```
Comment (eellison) [inline test/jit/test_freezing.py:1974]: is this needed if we're testing layernorm inplace above ?

**Review by eellison — APPROVED** [review]
LGTM, although i think the tests could be structured a bit better to be more readable

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:477]: fixed

File: test/jit/test_freezing.py (hunk 3 above)
Comment (Krovatkin) [reply test/jit/test_freezing.py:1961]: `len(param[1]) == 1` tests that we should only convert to `MKLDNNLayerNorm` iff the shape argument consists of a single dimension since this is what MKLDNN supports out of the box w/o collapsing the inner dimensions.

File: test/jit/test_freezing.py (hunk 4 above)
Comment (Krovatkin) [reply test/jit/test_freezing.py:1974]: I think so. I added `inplace` for mostly documentation purposes. We can't pass it directly to MKLDNN. We use it to create different graphs where we can use inplace (`Sequential`) and we cannot ( `ResidualLayerNorm`) since the input to the layernorm is also used *after* the layernorm so we can't inplace.

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:477]: is it ?

File: test/jit/test_freezing.py [hunk 5]
```diff
@@ -1939,7 +1950,42 @@ def test_layernorm(self):
...
+                        sub_model.eval()
+                        mod = torch.jit.freeze(torch.jit.script(sub_model))
+                        self.run_pass("convert_frozen_ops_to_mkldnn", mod.graph)
+                        # if weight and bias are present and shape is the last dimension
+                        # we should convert `aten::layer_norm` to `prim::MKLDNNLayerNorm`
+                        if layernorm.elementwise_affine and len(param[1]) == 1:
+                            inplace_suffix = "_" if inplace else ""
+                            FileCheck().check("prim::MKLDNNLayerNorm" + inplace_suffix).check_count("aten::to_dense", 1, exactly=True).run(mod.graph)
+                        else:
+                            FileCheck().check_count("aten::to_dense", 1, exactly=True).run(mod.graph)
```
Comment (eellison) [inline test/jit/test_freezing.py:1965]: this isnt testing that we didnt fuse layernorm

File: test/jit/test_freezing.py (hunk 4 above)
Comment (vadimkantorov) [reply test/jit/test_freezing.py:1974]: Could `F.layer_norm_`, `F.layer_norm(inplace = True)`, `nn.LayerNorm(inplace = True)` explicit variants be supported as well?
